src/lex.cpp

```diff
@@ -856,15 +857,22 @@ char8* lexer::parse_decimal_digits_and_underscores(char8* input) noexcept {
           input += 1;
         }
 
-        this->error_reporter_->report(
-            error_number_literal_contains_consecutive_underscores{
-                source_code_span(garbage_begin, input)});
+        if (is_digit(*input)) {
+          this->error_reporter_->report(
+              error_number_literal_contains_consecutive_underscores{
+                  source_code_span(garbage_begin, input)});
+        } else {
+          this->error_reporter_->report(
+            error_number_literal_contains_trailing_underscores{
+              source_code_span(garbage_begin, input)});
+        }
       }
     }
   }
-  if (has_trailing_underscore == true) {
-    // TODO error_reporter: SyntaxError: Numeric separators are not allowed at
-    // the end of numeric literals
+  if (garbage_begin != nullptr and has_trailing_underscore == true) {
```

Comment:
It looks like the MSVC compiler doesn't like `and`: https://github.com/strager/quick-lint-js/runs/1236834344

Let's use `&&` instead (or figure out how to make MSVC understand `and`).

Reply:
Normally it's corrected!
Congratulations for finding that MSVC does not like `and` in its (really verbose) logs.